dynamic_programming/coin_change_topdown.cpp

```diff
@@ -0,0 +1,89 @@
+/**
+ * @file
+ * @minimum coins change problem is a problem used to find the minimum number of 
+ * coins required to completely reach a target amount.
+ *
+ * @details
+ * This problem can be solved using 2 methods:
+ * 1. Top down approach
+ * 2. Bottom up appraoch
+ * Top down approach involves a vector with all elements initialised to 0.
+ * It is based on optimal substructure and overlapping subproblems.
+ * Overall time complexity of coin change problem is O(n*t)
+ * For example: example 1:-
+ * Coins: {1,7,10}
+ * Target:15
+ * Therfore minimum number of coins required = 3 of denomination 1,7 and 7.
+ * @author [Divyansh Kushwaha](https://github.com/webdesignbydivyansh)
+ * @see related problem at [Leetcode](https://leetcode.com/problems/coin-change/)
+*/
+
+#include <iostream>   // for io operations
+#include <vector>     // for std::vector
+#include <cassert>    // for assert
+#include <climits>    // for INT_MAX
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithm
+ */
+namespace dynamic_programming {
+/**
+ * @namespace mincoins_topdown
+ * @brief Functions for [minimum coin exchange](https://leetcode.com/problems/coin-change/) problem
+ */
+namespace mincoins_topdown {
+    /**
+    * @brief This implementation is for finding minimum number of coins .
+    * @param T is template type
+    * @param n amount to be reached
+    * @param coins vector of coins
+    * @param t deontes the number of coins
+    * @param dp initilised to 0
+    * @returns minimum number of coins
+    */
+    template<typename T>
+    int mincoins(T n, std::vector<T> coins, const int t, std::vector<T> dp){
```

Comment:
It is recommended/better to use `uint64_t`, rather than `int`.
```suggestion
    uint64_t mincoins(const T &n, const std::vector<T> &coins, const uint64_t &t, std::vector<T> dp){
```

Reply:
if i'm using uint64_t rather than int , the compiler gives segmentation fault. what do i do now?

Reply:
Try using `int64_t`, or another appropriate size for the variable ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂